ark/phenotyping/preprocess_test.py

```diff
@@ -0,0 +1,63 @@
+import os
+import pytest
+
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+import ark.phenotyping.preprocess as preprocess
+import ark.utils.misc_utils as misc_utils
+import ark.utils.test_utils as test_utils
+
+
+def test_create_pixel_matrix():
+    # must be float64, since that's how Candace's data comes in
+    sample_img_xr = test_utils.make_images_xarray(tif_data=None,
+                                                  fov_ids=['fov0', 'fov1'],
+                                                  channel_names=['chan0', 'chan1'],
+                                                  dtype='float64')
+
+    sample_labels = test_utils.make_labels_xarray(label_data=None,
+                                                  fov_ids=['fov0', 'fov1'],
+                                                  compartment_names=['whole_cell'])
+
+    # pass invalid fov names
+    with pytest.raises(ValueError):
+        preprocess.create_pixel_matrix(sample_img_xr, sample_labels, fovs=['fov1', 'fov2'])
+
+    # pass invalid channel names
+    with pytest.raises(ValueError):
+        preprocess.create_pixel_matrix(sample_img_xr, sample_labels, channels=['chan1', 'chan2'])
+
+    # test on all fovs and channels
+    flowsom_data = preprocess.create_pixel_matrix(sample_img_xr, sample_labels)
+
+    misc_utils.verify_same_elements(flowsom_fovs=flowsom_data['fov'].unique(),
+                                    provided_fovs=['fov0', 'fov1'])
+    misc_utils.verify_same_elements(flowsom_chans=flowsom_data.columns.values[:-4],
+                                    provided_chans=['chan0', 'chan1'])
+
+    # test on a subset of fovs
+    flowsom_data = preprocess.create_pixel_matrix(sample_img_xr, sample_labels, fovs=['fov1'])
+
+    misc_utils.verify_same_elements(flowsom_fovs=flowsom_data['fov'].unique(),
+                                    provided_fovs=['fov1'])
+    misc_utils.verify_same_elements(flowsom_chans=flowsom_data.columns.values[:-4],
+                                    provided_chans=['chan0', 'chan1'])
+
+    # test on a subset of channels
+    flowsom_data = preprocess.create_pixel_matrix(sample_img_xr, sample_labels, channels=['chan0'])
+
+    misc_utils.verify_same_elements(flowsom_fovs=flowsom_data['fov'].unique(),
+                                    provided_fovs=['fov0', 'fov1'])
+    misc_utils.verify_same_elements(flowsom_chans=flowsom_data.columns.values[:-4],
+                                    provided_chans=['chan0'])
+
+    # test on a subset of fovs and channels
+    flowsom_data = preprocess.create_pixel_matrix(sample_img_xr, sample_labels,
+                                                  fovs=['fov1'], channels=['chan0'])
+
+    misc_utils.verify_same_elements(flowsom_fovs=flowsom_data['fov'].unique(),
+                                    provided_fovs=['fov1'])
+    misc_utils.verify_same_elements(flowsom_chans=flowsom_data.columns.values[:-4],
```

Comment:
Some light correctness checks:
Make sure there are no blank rows in the output
Make sure that all of the rows sum to 1 (this will depend on when the 99.9% normalization happens, see above)
Generate five random row + col indices, extract the marker counts from those indices, convert to frequency, and check that they match the corresponding row in the output.

Reply:
@ngreenwald does making sure that all the rows sum to 1 cover the case with no blank rows, or are you thinking of something different (ex. NaN does not appear)?

Reply:
Good point, this check actually covers both cases

Reply:
@ngreenwald since 99.9% normalization happens after converting to frequency, I don't think the rows of the final dataframe will sum to 1 anymore.